lib/src/generator/file_structure.dart

```diff
@@ -20,22 +18,76 @@ enum FileStructureMode {
 /// to lay them out on disk, and how to link different pages in the structure
 /// together.
 abstract class FileStructure {
-  factory FileStructure(FileStructureMode mode, ModelElement modelElement) {
-    switch (mode) {
-      case FileStructureMode.htmlOriginal:
-        return _FileStructureHtml(modelElement);
-      case FileStructureMode.mdOriginal:
-        return _FileStructureMd(modelElement);
+  factory FileStructure.fromDocumentable(Documentable documentable) {
+    if (!_validFormats.contains(documentable.config.format)) {
+      throw DartdocFailure(
+          'Internal error: unrecognized config.format: ${documentable.config.format}');
+    }
+    switch (documentable) {
+      case LibraryContainer():
+        // [LibraryContainer]s are not ModelElements, but have documentation.
+        return FileStructure._fromLibraryContainer(documentable);
+      case ModelElement():
+        // This should be the common case.
+        return FileStructure._fromModelElement(documentable);
+      default:
+        throw UnimplementedError(
+            'Tried to build a FileStructure for an unknown subtype of Documentable:  ${documentable.runtimeType}');
     }
   }
 
-  /// Link to the [ModelElement] the information for this [FileStructure]
-  /// applies to.
-  // TODO(jcollins): consider not carrying a reference to a ModelElement and
-  // calculating necessary bits at construction time.  Might be challenging
-  // if we want to calculate [hasIndependentFile] based on documentation
-  // length or other variables not always available.
-  ModelElement get modelElement;
+  factory FileStructure._fromLibraryContainer(
+      LibraryContainer libraryContainer) {
+    String? kindAddition;
+    String? pathSafeName = libraryContainer.name;
+    switch (libraryContainer) {
+      case Category():
+        kindAddition = 'topic';
+        break;
+      case Package():
+        pathSafeName = 'index';
+        break;
+      default:
+        throw UnimplementedError(
+            'Unrecognized LibraryContainer subtype:  ${libraryContainer.runtimeType}');
+    }
+
+    return FileStructureImpl(
```

Comment:
It seems to me a lot of ceremony, to declare `kindAddition`, and `pathSafeName`, maybe re-assign to them, and then just return a new FileStructureImpl. Up to you but it could just be:

```dart
switch (libraryContainer) {
  case Category():
    return FileStructureImpl(
        libraryContainer.config.format, libraryContainer.name, 'topic');
  case Package():
    return FileStructureImpl(
        libraryContainer.config.format, 'index', null);
  default:
    throw ...
}
```

Reply:
Hmmm.  I think the fact I rewrote this several times is leaking through.  Yes, that's much better.   Fixing here and in _fromModelElement.